packages/compartment-mapper/src/import-hook.js

```diff
@@ -398,3 +405,256 @@ export const makeImportHookMaker = (
+          packageLocation,
+        );
+        // eslint-disable-next-line no-await-in-loop
+        /** @type {Uint8Array} */
+        let moduleBytes;
+        try {
+          moduleBytes = readSync(moduleLocation);
+        } catch (err) {
+          if (err && err.code === 'ENOENT') {
+            // might be an exit module. use the fallback `exitModuleImportNowHook` to import it
+            // eslint-disable-next-line no-continue
+            continue;
+          }
+          throw err;
+        }
+        if (moduleBytes !== undefined) {
+          /** @type {string | undefined} */
+          let sourceMap;
+          // eslint-disable-next-line no-await-in-loop
+          const envelope = parse(
+            moduleBytes,
+            candidateSpecifier,
+            moduleLocation,
+            packageLocation,
+            {
+              compartmentDescriptor,
+              readPowers,
+              sourceMapHook:
+                sourceMapHook &&
+                (nextSourceMapObject => {
+                  sourceMap = JSON.stringify(nextSourceMapObject);
+                }),
+            },
+          );
+          const {
+            parser,
+            bytes: transformedBytes,
+            record: concreteRecord,
+          } = envelope;
+
+          // Facilitate a redirect if the returned record has a different
+          // module specifier than the requested one.
+          if (candidateSpecifier !== moduleSpecifier) {
+            moduleDescriptors[moduleSpecifier] = {
+              module: candidateSpecifier,
+              compartment: packageLocation,
+            };
+          }
+          /** @type {StaticModuleType} */
+          const record = {
+            record: concreteRecord,
+            specifier: candidateSpecifier,
+            importMeta: { url: moduleLocation },
+          };
+
+          let sha512;
+          if (computeSha512 !== undefined) {
+            sha512 = computeSha512(transformedBytes);
+
+            if (sourceMapHook !== undefined && sourceMap !== undefined) {
+              sourceMapHook(sourceMap, {
+                compartment: packageLocation,
+                module: candidateSpecifier,
+                location: moduleLocation,
+                sha512,
+              });
+            }
+          }
+
+          const packageRelativeLocation = moduleLocation.slice(
+            packageLocation.length,
+          );
+          packageSources[candidateSpecifier] = {
+            location: packageRelativeLocation,
+            sourceLocation: moduleLocation,
+            parser,
+            bytes: transformedBytes,
+            record: concreteRecord,
+            sha512,
+          };
+          for (const importSpecifier of getImportsFromRecord(record)) {
+            strictlyRequiredForCompartment(packageLocation).add(
+              resolve(importSpecifier, moduleSpecifier),
+            );
+          }
+
+          return record;
+        }
+      }
```

Comment:
copypasta

Reply:
I think we can bounce most of it out https://github.com/endojs/endo/commit/afa4e6354aa79de93307fbfd3eb8ea443a9e6e7f